src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -77,48 +113,72 @@ private void MakeInterpolatedStringFormat(BoundInterpolatedString node, out Boun
-            //
-            //     String.Format("Jenny don\'t change your number {0}", new object[] { 8675309 })
-            //
-
             Debug.Assert(node.Type.SpecialType == SpecialType.System_String); // if target-converted, we should not get here.
-            BoundExpression format;
-            ArrayBuilder<BoundExpression> expressions;
-            MakeInterpolatedStringFormat(node, out format, out expressions);
-            if (expressions.Count == 0)
+
+            BoundExpression result;
+
+            if (CanLowerToStringConcatenation(node))
             {
-                // There are no fill-ins. Handle the escaping of {{ and }} and return the value.
-                Debug.Assert(!format.HasErrors && format.ConstantValue != null && format.ConstantValue.IsString);
-                var builder = PooledStringBuilder.GetInstance();
-                var formatText = format.ConstantValue.StringValue;
-                int formatLength = formatText.Length;
-                for (int i = 0; i < formatLength; i++)
+                // All fill-ins, if any, are strings, and none of them have alignment or format specifiers.
+                // We can lower to a more efficient string concatenation
+                // The normal pattern for lowering is to lower subtrees before the enclosing tree. However in this case
+                // we want to lower the entire concatenation so we get the optimizations done by that lowering (e.g. constant folding).
+
+                int length = node.Parts.Length;
+                if (length == 0)
+                {
+                    // $"" -> ""
+                    return _factory.StringLiteral("");
+                }
+
+                result = null;
+                for (int i = 0; i < length; i++)
                 {
-                    char c = formatText[i];
-                    builder.Builder.Append(c);
-                    if ((c == '{' || c == '}') && (i + 1) < formatLength && formatText[i + 1] == c)
+                    var part = node.Parts[i];
+                    if (part is BoundStringInsert fillin)
                     {
-                        i++;
+                        // this is one of the filled-in expressions
+                        part = fillin.Value;
                     }
+                    else
+                    {
+                        // this is one of the literal parts
+                        Debug.Assert(part is BoundLiteral && part.ConstantValue != null);
+                        part = _factory.StringLiteral(Unescape(part.ConstantValue.StringValue));
+                    }
+
+                    result = result == null ?
+                        part :
+                        _factory.Binary(BinaryOperatorKind.StringConcatenation, node.Type, result, part);
                 }
-                return _factory.StringLiteral(builder.ToStringAndFree());
+
+                result = _factory.Coalesce(result, _factory.StringLiteral(""));
+            }
```

Comment:
I don't think we should do this unless there was no string concatenation.

Reply:
It is not functionally wrong to do this, even if there was a concatenation, because we remove it again in the lowering of the coalescing operator. I noticed this technique is used in lot's of places. (For example, the lowering of a non-binary string concatenation builds lot's of intermediary structures that are broken down again when lowering a higher level.) So I used the same technique here to clean up the code considerably. Also, notice that the unit tests all still pass, nothing has changed there.

But you are correct, a simple `if (length == 1)` avoids it in most cases.